source/geod24/LocalRest.d

```diff
@@ -72,6 +81,101 @@ private struct ArgWrapper (T...)
     T args;
 }
 
+/// Our own little scheduler
+private final class LocalScheduler : C.FiberScheduler
+{
+    import core.sync.condition;
+    import core.sync.mutex;
+
+    /// Just a FiberCondition with a state
+    private struct Waiting { FiberCondition c; bool busy; }
+
+    /// The 'Response' we are currently processing, if any
+    private Response pending;
+    /// List of Condition for blocked queries
+    /// Some entries might be empty (we never size it down)
+    private Waiting[] waiting;
+
+    /// Should never be called from outside
+    public override Condition newCondition(Mutex m = null) nothrow
+    {
+        assert(0);
+    }
+
+    /// Get the next available request ID
+    public size_t getNextResponseId ()
+    {
+        // Try to find one in the array
+        foreach (idx, ref val; this.waiting)
+            if (!val.busy)
+                return idx;
+        return this.waiting.length;
+    }
+
+    public Response waitResponse (size_t id) nothrow
+    {
+        if (id == this.waiting.length)
+            this.waiting ~= Waiting(new FiberCondition, false);
+        else if (id > this.waiting.length)
+            assert(0, "This should never happend");
+
+        Waiting* ptr = &this.waiting[id];
+        if (ptr.busy)
+            assert(0, "Trying to override a pending request");
+
+        // We yield and wait for an answer
+        ptr.busy = true;
+        ptr.c.wait();
+        ptr.busy = false;
+        // After control returns to us, `pending` has been filled
+        scope(exit) this.pending = Response.init;
+        return this.pending;
+    }
+
+    /// Override `FiberScheduler.FiberCondition` to avoid mutexes
+    /// and usage of global state
+    private class FiberCondition : Condition
+    {
+        this() nothrow
+        {
+            super(null);
+            notified = false;
+        }
+
+        override void wait() nothrow
+        {
+            scope (exit) notified = false;
+            while (!notified)
+                this.outer.yield();
+        }
+
+        override bool wait(Duration period) nothrow
+        {
+            assert(0); // Unused
+        }
+
+        override void notify() nothrow
+        {
+            notified = true;
+            this.outer.yield();
+        }
+
+        override void notifyAll() nothrow
+        {
+            notified = true;
+            this.outer.yield();
+        }
+
+        private bool notified;
+    }
+}
+
+
+/// We need a scheduler to simulate an event loop and to be re-entrant
+/// However, the one in `std.concurrency` is process-global (`__gshared`)
+private LocalScheduler scheduler;
```

Comment:
It's a bit ugly to have it in module scope, but I guess it's complicated to move it inside RemoteAPI because we would have multiple instances of it for every type RemoteAPI is instantiated with..

Reply:
It is ugly but I don't see any way around it.